### PR TITLE
=str #25469 stream refs hardened termination via awaiting completion signal

### PR DESCRIPTION
**Pull request by ktoso**
Resolves https://github.com/akka/akka/issues/25469

However I want to test and ponder it some more... 

I do think the scheduling dance should be on that receiving side though, since on the sending side we could delay termination as well allowing more time for the completion/failure to be sent over, however it feels more right for the receiver to be able to tweak how tolerant they are to these things.

### CODE REVIEW

File: akka-stream/src/main/resources/reference.conf [hunk 1]
```diff
@@ -109,6 +109,9 @@ akka {
         # prepare things before it is ready to materialize the reference. However the timeout is needed to avoid leaking
         # in-active streams which are never subscribed to.
         subscription-timeout = 30 seconds
+
+        #
+        backing-actor-termination-delay = 2 seconds
```
Comment (ktoso) [inline akka-stream/src/main/resources/reference.conf]: crappy name, will rename and add docs

File: akka-stream/src/main/resources/reference.conf [hunk 2]
```diff
@@ -109,6 +109,9 @@ akka {
         # prepare things before it is ready to materialize the reference. However the timeout is needed to avoid leaking
         # in-active streams which are never subscribed to.
         subscription-timeout = 30 seconds
+
+        #
```
Comment (johanandren) [inline akka-stream/src/main/resources/reference.conf]: Better description!

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/StreamRefsSpec.scala [hunk 3]
```diff
@@ -164,11 +164,22 @@ object StreamRefsSpec {
...
+           enabled = on
+           transport = aeron-udp
+           canonical.hostname = "${address.getHostName}"
+           canonical.port =  ${address.getPort}
+           log-received-messages = on
+           log-sent-messages = on
+         }
+       }
+      stream.materializer.stream-ref {
+        subscription-timeout = 3 seconds
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/StreamRefsSpec.scala]: Can it be covered by a test case by manually sending it a Terminated?

**Review by johanandren — COMMENTED** [review]
Lookin' good.

File: akka-stream/src/main/resources/reference.conf (hunk 2 above)
Comment (johanandren) [reply akka-stream/src/main/resources/reference.conf]: Ah, saw your own comment now! NVM

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/StreamRefsSpec.scala (hunk 3 above)
Comment (ktoso) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/StreamRefsSpec.scala]: Getting the right actor ref is a bit tricky... will try tho

File: akka-stream/src/main/resources/reference.conf (hunk 2 above)
Comment (ktoso) [reply akka-stream/src/main/resources/reference.conf]: Yes, as I mentioned I would do: https://github.com/akka/akka/pull/25561/files#r214632185 :)

**Review by patriknw — COMMENTED** [review]
Looking good, probably difficult to write a deterministic test. Perhaps write one that would sometimes fail without the bug fix.

**Comment by drexin** [comment]
Thanks for the fix. I think for most cases this would work, but I wonder if the scheduled timeout is necessary. Wouldn't it be sufficient to have the downstream/receiver handle the completion? That way we wouldn't have a race and in case of an actual failure we wouldn't have a delay. I see following scenarios:

1.) Everything goes well

- Sender sends all messages followed by completion message and goes into completed state, but doesn't shut down, instead waits for `Terminated` from receiver
- Receiver receives all messages including completion message and shuts itself down
- Sender receives `Terminated` and shuts itself down

2.) Sender fails with exception

- Same as in successful case, only the message is different

3.) Downstream fails or cancels, one of the nodes fails or net split

- Should be handled by deathwatch

Am I missing anything?

**Comment by ktoso** [comment]
The above 3 cases missed a 4th one: we get a cluster partition and down nodes; then we need to stop the other sides instead of having them linger forever - would be a kind of memory leak really. So we need to react, on a timeout to a Terminated if no other signal arrives at all later on as well -- so the timer is unavoidable. Though your proposal @drexin hardens it well in the more normal cases, so I'll get us a hybrid version here.

**Comment by ktoso** [comment]
Nope that's wrong as well... something else must be causing the issue I'm seeing... Since we delay termination of sender, the receiver should not get that terminated but if it DOES then it means it was a partition and it will never get the final complete/fail message and must terminate as soon as it sees that terminated... So that fits the above model...

File: akka-stream/src/main/resources/reference.conf [hunk 4]
```diff
@@ -109,6 +109,14 @@ akka {
...
         # in-active streams which are never subscribed to.
         subscription-timeout = 30 seconds
+
+        # On transports that use separate lanes for system messages (such as Terminated) the completion signal from a
+        # SourceRef may reach the recieving end _after_ the Terminated message which may cause seeminly spurious failures
+        # of a stream. In order to avoid this, if this race happens, more time is given for the normal completion signal
+        # to arrive (set by this configuration option). If no proper completion signal arrives in that time, we assume
+        # that the other side has terminated not-gracefully (e.g. the node was marked as Down), and signal a failure on
+        # the receiving side of the stream.
+        termination-received-before-completion-leeway = 2 seconds
```
Comment (ktoso) [inline akka-stream/src/main/resources/reference.conf]: whoops forgot to rename

**Review by johanandren — COMMENTED** [review]
Left some questions and thoughts.

File: akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala [hunk 5]
```diff
@@ -136,7 +148,8 @@ private[stream] final class SinkRefStageImpl[In] private[akka] (
             s"[$stageActorName] Remote side did not subscribe (materialize) handed out Source reference [${promise.future.value}], " +
               s"within subscription timeout: ${PrettyDuration.format(subscriptionTimeout.timeout)}!")
 
-          throw ex // this will also log the exception, unlike failStage; this should fail rarely, but would be good to have it "loud"
+          getPartnerRef ! StreamRefsProtocol.RemoteStreamFailure(ex.getMessage)
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala]: is partnerRef always known when SubscriptionTimeout?

File: akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala [hunk 6]
```diff
@@ -136,7 +148,8 @@ private[stream] final class SinkRefStageImpl[In] private[akka] (
...
+          finishedWithAwaitingPartnerTermination = OptionVal(Failure(ex))
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala]: what is triggering the final tear down in this SubscriptionTimeout case?
We act on the `finishedWithAwaitingPartnerTermination` when receiving `Terminated`, but we only watch after receiving the partner. How is that related to the subscription timeout?

File: akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala [hunk 7]
```diff
@@ -145,26 +158,29 @@ private[stream] final class SinkRefStageImpl[In] private[akka] (
...
 
       override def onUpstreamFinish(): Unit =
         partnerRef match {
           case OptionVal.Some(ref) ⇒
             ref ! StreamRefsProtocol.RemoteStreamCompleted(remoteCumulativeDemandConsumed)
-            self.unwatch(getPartnerRef)
-            super.onUpstreamFinish()
+            finishedWithAwaitingPartnerTermination = OptionVal(Success(Done))
+            setKeepGoing(true) // we will terminate once partner ref has Terminated (to avoid racing Terminated with completion message)
+          // super.onUpstreamFailure()
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala]: remove

File: akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala [hunk 8]
```diff
@@ -145,26 +158,29 @@ private[stream] final class SinkRefStageImpl[In] private[akka] (
...
-      override def onUpstreamFailure(ex: Throwable): Unit =
+      override def onUpstreamFailure(ex: Throwable): Unit = {
         partnerRef match {
           case OptionVal.Some(ref) ⇒
             ref ! StreamRefsProtocol.RemoteStreamFailure(ex.getMessage)
-            self.unwatch(getPartnerRef)
-            super.onUpstreamFailure(ex)
+            finishedWithAwaitingPartnerTermination = OptionVal(Failure(ex))
+            setKeepGoing(true) // we will terminate once partner ref has Terminated (to avoid racing Terminated with completion message)
+          // super.onUpstreamFailure(ex)
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala]: remove

File: akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala [hunk 9]
```diff
@@ -181,14 +197,16 @@ private[stream] final class SinkRefStageImpl[In] private[akka] (
 
           completedBeforeRemoteConnected match {
             case OptionVal.Some(scala.util.Failure(ex)) ⇒
-              log.warning("Stream already terminated with exception before remote side materialized, failing now.")
+              log.warning("Stream already terminated with exception before remote side materialized, sending failure: " + ex.getMessage)
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala]: use logging placeholders instead of string concatenation.
`toString` of Exception is often better than `getMessage`, includes exception class

File: akka-stream/src/main/scala/akka/stream/impl/streamref/SourceRefImpl.scala [hunk 10]
```diff
@@ -138,6 +139,11 @@ private[stream] final class SourceRefStageImpl[Out](
           log.debug("[{}] Scheduled re-delivery of demand until [{}]", stageActorName, localCumulativeDemand)
           getPartnerRef ! StreamRefsProtocol.CumulativeDemand(localCumulativeDemand)
           scheduleDemandRedelivery()
+
+        case TerminationDeadlineTimerKey ⇒
+          log.warning("[{}] Failing since termination deadline exceeded, no final completion/failure message was received from other side within {}", stageActorName, settings.finalTerminationSignalDeadline)
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/streamref/SourceRefImpl.scala]: pretty duration

**Review by patriknw — COMMENTED** [review]
looking good, some questions around subscription timeout

File: akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala (hunk 9 above)
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala]: Oh, also, don't we have a `warning(ex, message)` nowadays (or is that only in typed)?

File: akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala [hunk 11]
```diff
@@ -79,6 +79,11 @@ private[stream] final class SinkRefStageImpl[In] private[akka] (
 
       private var completedBeforeRemoteConnected: OptionVal[Try[Done]] = OptionVal.None
 
+      // `true` when this side of the stream has completed/failed, and we await the Terminated() signal back from the partner
+      // so we can safely shut down completely; This is to avoid *our* Terminated() signal to reach the partner before the
+      // Complete/Fail message does, which can happen on transports such as Artery which use a dedicated lane for system messages (Terminated)
+      private[this] var finishedWithAwaitingPartnerTermination: OptionVal[Try[Done]] = OptionVal.None
```
[the remark this answers is not in the document]
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala]: Whoops thanks, used to be a boolean

File: akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala (hunk 5 above)
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala]: Thanks good catch, this is wrong -- we never will have the ref here; the timeout triggers when we did NOT get the ref after all.

File: akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala]: Thanks yeah it was the wrong way around; this should remain as throwing

Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala]: this was a real bug, thanks for finding -- it has to throw

File: akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala (hunk 9 above)
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala]: The overload is only Typed @johanandren :)
Thanks, changed

File: akka-stream/src/main/scala/akka/stream/impl/streamref/SourceRefImpl.scala [hunk 12]
```diff
@@ -138,6 +139,11 @@ private[stream] final class SourceRefStageImpl[Out](
...
+          failStage(RemoteStreamRefActorTerminatedException(s"Remote partner [$partnerRef] has terminated unexpectedly and no clean completion/failure message was received " +
+            "(possible reasons: network partition or subscription timeout triggered termination of partner). Tearing down."))
```
[the remark this answers is not in the document]
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/impl/streamref/SourceRefImpl.scala]: Ok, I'll only fail here.

File: akka-stream/src/main/scala/akka/stream/impl/streamref/SourceRefImpl.scala [hunk 13]
```diff
@@ -173,21 +179,17 @@ private[stream] final class SourceRefStageImpl[Out](
...
-            case _ ⇒
-              // this should not have happened! It should be impossible that we watched some other actor
-              failStage(RemoteStreamRefActorTerminatedException(s"Received UNEXPECTED Terminated($ref) message! " +
-                s"This actor was NOT our trusted remote partner, which was: $getPartnerRef. Tearing down."))
-
-          }
+          // we need to start a delayed shutdown in case we were network partitioned and the final signal complete/fail
+          // will never reach us; so after the given timeout we need to forcefully terminate this side of the stream ref
+          // the other (sending) side terminates by default once it gets a Terminated signal so no special handling is needed there.
+          scheduleOnce(TerminationDeadlineTimerKey, settings.finalTerminationSignalDeadline)
```
[the remark this answers is not in the document]
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/impl/streamref/SourceRefImpl.scala]: Actually I will bring back that early failing... hmm it is more correct and could uncover a bug to fail eagerly here if it ever happened

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by ktoso** [comment]
Thanks, 2nd review @johanandren please?

**Review by johanandren — APPROVED** [review]
